### DISCUSSION TITLE
Uploading hex string as bytes

### DISCUSSION THREAD

**Discussion by dino-rodriguez** [foundry-rs/foundry#2621] [Help] [answered]
I'm currently trying to cast a hex string in forge to bytes and uploading it. But the string `0xaa`, for example, gets uploaded as 4 bytes when it should be 1 byte. This is the approach I am taking:
```
bytes[] public traits = [bytes(<in memory string>)]
```

In hardhat this works "magically". You can pass a hex string to a solidity function that accepts bytes, and it does the right thing. e.g. `0xaa` is one byte.

How do I successfully upload hex as bytes using forge?

**Comment by onbjerg** [comment, marked as answer]
This is a Solidity question for future reference, so you should be able to find the answer to that in the Solidity docs.

I think what you're looking for is `hex""`, e.g. `hex"aa"`

**Reply by dino-rodriguez** [reply]
Sorry, I didn't give a great example, edited it above to reflect what I am trying to do. I am aware I can do that with string literals. I am trying to do it with strings loaded in memory from a file using the foundry cheatcode. I don't think I can use `hex` for anything but string literals.

**Reply by onbjerg** [reply]
If you're using FFI and the output is well-formed hex starting with `0x` it should give you back bytes, not a string

**Reply by dino-rodriguez** [reply]
Great, thank you. I'll give that a shot.
